dom/payments/MerchantValidationEvent.cpp

```diff
@@ -0,0 +1,188 @@
+/* -*- Mode: C++; tab-width: 8; indent-tabs-mode: nil; c-basic-offset: 2 -*- */
+/* vim: set ts=8 sts=2 et sw=2 tw=80: */
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#include "mozilla/dom/MerchantValidationEvent.h"
+#include "nsIURLParser.h"
+#include "nsNetCID.h"
+#include "mozilla/dom/PaymentRequest.h"
+#include "mozilla/dom/Location.h"
+
+namespace mozilla {
+namespace dom {
+
+NS_IMPL_CYCLE_COLLECTION_INHERITED(MerchantValidationEvent, Event, mRequest)
+
+NS_IMPL_CYCLE_COLLECTION_TRACE_BEGIN_INHERITED(MerchantValidationEvent, Event)
+NS_IMPL_CYCLE_COLLECTION_TRACE_END
+
+NS_INTERFACE_MAP_BEGIN_CYCLE_COLLECTION(MerchantValidationEvent)
+NS_INTERFACE_MAP_END_INHERITING(Event)
+
+NS_IMPL_ADDREF_INHERITED(MerchantValidationEvent, Event)
+NS_IMPL_RELEASE_INHERITED(MerchantValidationEvent, Event)
+
+class Promise;
+class EventTarget;
+
+already_AddRefed<MerchantValidationEvent>
+MerchantValidationEvent::Constructor(
+  EventTarget* aOwner,
+  const nsAString& aType,
+  const MerchantValidationEventInit& aEventInitDict)
+{
+  RefPtr<MerchantValidationEvent> e = new MerchantValidationEvent(aOwner);
+  bool trusted = e->Init(aOwner);
+  e->InitEvent(aType, aEventInitDict.mBubbles, aEventInitDict.mCancelable);
+  e->SetTrusted(trusted);
+  e->SetComposed(aEventInitDict.mComposed);
+  return e.forget();
+}
+
+already_AddRefed<MerchantValidationEvent>
+MerchantValidationEvent::Constructor(
+  const GlobalObject& aGlobal,
+  const nsAString& aType,
+  const MerchantValidationEventInit& aEventInitDict,
+  ErrorResult& aRv)
+{
+  nsCOMPtr<EventTarget> owner = do_QueryInterface(aGlobal.GetAsSupports());
+  auto event = Constructor(owner, aType, aEventInitDict);
+  // Let base be the event’s relevant settings object’s API base URL.
+  nsCOMPtr<nsPIDOMWindowInner> window = do_QueryInterface(owner);
+  RefPtr<Location> location = window->GetLocation();
+  // TODO: ENSURE LOCATION
```

Comment:
Can I do the following here?:

```cpp
auto location = window->GetLocation();
MOZ_ASSERT(location);
```

Reply:
I think it should be 
  nsCOMPtr<nsIDocument> doc = window->GetExtantDoc();
  nsCOMPtr<nsIPrincipal> principal = doc->NodePrincipal();
then you can get uri/domain information from nsIPrincipal
https://searchfox.org/mozilla-central/source/__GENERATED__/dist/include/nsIPrincipal.h#60